src/index.ts

```diff
@@ -0,0 +1,121 @@
+import { type RemotePinningServiceClient, type Pin, type PinStatus, type PinsRequestidPostRequest, Status } from '@ipfs-shipyard/pinning-service-client'
+import { multiaddr } from '@multiformats/multiaddr'
+import debug from 'debug'
+import pRetry from 'p-retry'
+import { FailedToConnectToDelegates } from './errors.js'
+import type { Helia } from '@helia/interface'
+import type { CID } from 'multiformats/cid'
+
+const log = debug('helia-remote-pinning')
+const logError = log.extend('error')
+const logTrace = log.extend('trace')
+
+export interface HeliaRemotePinningOptions {
+  /**
+   * Control whether requests are aborted or not by manually aborting a signal or using AbortSignal.timeout()
+   */
+  signal?: AbortSignal
+
+  /**
+   * The CID instance to pin. When using Helia, passing around the CID object is preferred over the string.
+   */
+  cid: CID
+}
+
+export interface AddPinArgs extends Omit<Pin, 'cid'>, HeliaRemotePinningOptions {}
+
+export interface ReplacePinArgs extends Omit<PinsRequestidPostRequest, 'pin'>, Omit<Pin, 'cid'>, HeliaRemotePinningOptions {}
+
+export class HeliaRemotePinner {
+  constructor (private readonly heliaInstance: Helia, private readonly remotePinningClient: RemotePinningServiceClient) {
+  }
+
+  private async getOrigins (otherOrigins: Pin['origins']): Promise<Set<string>> {
+    const origins = new Set(this.heliaInstance.libp2p.getMultiaddrs().map(multiaddr => multiaddr.toString()))
+    if (otherOrigins != null) {
+      for (const origin of otherOrigins) {
+        origins.add(origin)
+      }
+    }
+    return origins
+  }
+
+  private async connectToDelegates (delegates: Set<string>, signal?: AbortSignal): Promise<void> {
+    let successfulDials = 0
+    try {
+      for (const delegate of delegates) {
+        await this.heliaInstance.libp2p.dial(multiaddr(delegate), { signal })
+        successfulDials++
+      }
```

Comment:
Maybe use `Promise.any` to dial in parallel?  Then one slow delegate won't make things unresponsive.